[logging] Move population of common MetricsContext fields to record_compilation_metrics

masnesral opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #141291

Summary: Fix outstanding TODOs related to logging of CompilationMetrics by moving the population of common fields to record_compilation_metrics() instead of populating those independently wherever we use a the metrics_context contextmanager:
* Keep track of start and end time in MetricsContext and pass those to record_compilation_metrics() and populate those fields in that function.
* Pass exception info to record_compilation_metrics() and populate those field in that function.
* Add a new contextmanager, chromium_event_timed, to create the start/end "dynamo" event. This is important because I want this contextmanager to complete _after_ building the CompilationMetrics.
* Populate the compile_id field centrally in record_compilation_metrics().
* Populate the structured_logging_overhead centrally in record_compilation_metrics().
* Add the CompilationMetrics to the current chromium event in record_compilation_metrics(), after all common fields have been added. In a future diff, I can also add _all_ compilation metrics to the chromium event.

Test plan: Unit tests. Also see internal testing:
* dynamo_compile: https://fburl.com/scuba/dynamo_compile/sandbox/jrascnf9
* pt2_compile_events: https://fburl.com/scuba/pt2_compile_events/l3jnla06
* tlparse: https://fburl.com/bq5a9nqs

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

masnesral committed:
[logging] Move population of common MetricsContext fields to record_compilation_metrics

Summary: Fix outstanding TODOs related to logging of CompilationMetrics by moving the population of common fields to record_compilation_metrics() instead of populating those independently wherever we use a the metrics_context contextmanager:
* Keep track of start and end time in MetricsContext and pass those to record_compilation_metrics() and populate those fields in that function.
* Pass exception info to record_compilation_metrics() and populate those field in that function.
* Add a new contextmanager, chromium_event_timed, to create the start/end "dynamo" event. This is important because I want this contextmanager to complete _after_ building the CompilationMetrics.
* Populate the compile_id field centrally in record_compilation_metrics().
* Populate the structured_logging_overhead centrally in record_compilation_metrics().
* Add the CompilationMetrics to the current chromium event in record_compilation_metrics(), after all common fields have been added. In a future diff, I can also add _all_ compilation metrics to the chromium event.

Test plan: Unit tests. Also see internal tlparse here: https://fburl.com/hv3inju2

[ghstack-poisoned]

masnesral committed:
Update on "[logging] Move population of common MetricsContext fields to record_compilation_metrics"


Summary: Fix outstanding TODOs related to logging of CompilationMetrics by moving the population of common fields to record_compilation_metrics() instead of populating those independently wherever we use a the metrics_context contextmanager:
* Keep track of start and end time in MetricsContext and pass those to record_compilation_metrics() and populate those fields in that function.
* Pass exception info to record_compilation_metrics() and populate those field in that function.
* Add a new contextmanager, chromium_event_timed, to create the start/end "dynamo" event. This is important because I want this contextmanager to complete _after_ building the CompilationMetrics.
* Populate the compile_id field centrally in record_compilation_metrics().
* Populate the structured_logging_overhead centrally in record_compilation_metrics().
* Add the CompilationMetrics to the current chromium event in record_compilation_metrics(), after all common fields have been added. In a future diff, I can also add _all_ compilation metrics to the chromium event.

Test plan: Unit tests. Also see internal tlparse here: https://fburl.com/hv3inju2

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

masnesral commented on torch/_dynamo/convert_frame.py:
```diff
@@ -873,17 +870,11 @@ def count_args(code: CodeType) -> int:
 
         return guarded_code
 
-    chromium_event_log = get_chromium_event_logger()
```
Replaced with a new contextmanager.

masnesral commented on torch/_dynamo/convert_frame.py:
```diff
@@ -1120,7 +1101,6 @@ def clean_for_json(d: Dict[str, Any]) -> Dict[str, Any]:
 
             config_dict = clean_for_json(config.get_config_copy())
             metrics = {
-                "compile_id": str(compile_id),
```
All these fields are common to the logging for the backwards, so they're all moving to `record_compilation_metrics`

masnesral commented on torch/_dynamo/utils.py:
```diff
@@ -1067,6 +1093,13 @@ def get_stack(self):
             self.tls.stack = []
             return self.tls.stack
 
+    def get_top(self) -> str:
```
@jamesjwu This and the contextmanager below is probably the part where I need your input the most.

masnesral commented on torch/_functorch/_aot_autograd/runtime_wrappers.py:
```diff
@@ -2019,53 +2018,28 @@ def _backward_impl(ctx, all_args):
-                            fail_type = type(e).__qualname__
-                            fail_reason = str(e)
-                            raise
-                        finally:
```
All this crap moves to `record_compilation_metrics`

jamesjwu commented on torch/_dynamo/utils.py:
```diff
@@ -1300,6 +1333,32 @@ def get_chromium_event_logger() -> ChromiumEventLogger:
+        yield
+    finally:
+        chromium_event_log.log_event_end(
+            event_name, time.time_ns(), {}, chromium_start_time, True
```
Don't think this should be always True, should be `log_pt2_compile_event`.

jamesjwu commented:
Other than my other comment, this looks fine overall:  

Can you also test the logging to our internal table? To do so, run a benchmark internally using buck, but also turn on `TORCH_LOGS=dynamo`. 

You should see a line like:

```
ChromiumEventLogger initialized with id <id>
```

Take that id and look at the internal PT2 Compile Events table. Find all the events and make sure that the events seem reasonable and expected. Make sure that the `unknown_columns` column is empty. You can check the `event=dynamo` in Scuba to make sure those events look "fine" (or just link it here and I can take a look)!

jamesjwu commented:
LGTM

masnesral committed:
Update on "[logging] Move population of common MetricsContext fields to record_compilation_metrics"


Summary: Fix outstanding TODOs related to logging of CompilationMetrics by moving the population of common fields to record_compilation_metrics() instead of populating those independently wherever we use a the metrics_context contextmanager:
* Keep track of start and end time in MetricsContext and pass those to record_compilation_metrics() and populate those fields in that function.
* Pass exception info to record_compilation_metrics() and populate those field in that function.
* Add a new contextmanager, chromium_event_timed, to create the start/end "dynamo" event. This is important because I want this contextmanager to complete _after_ building the CompilationMetrics.
* Populate the compile_id field centrally in record_compilation_metrics().
* Populate the structured_logging_overhead centrally in record_compilation_metrics().
* Add the CompilationMetrics to the current chromium event in record_compilation_metrics(), after all common fields have been added. In a future diff, I can also add _all_ compilation metrics to the chromium event.

Test plan: Unit tests. Also see internal tlparse here: https://fburl.com/hv3inju2

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

masnesral committed:
Update on "[logging] Move population of common MetricsContext fields to record_compilation_metrics"


Summary: Fix outstanding TODOs related to logging of CompilationMetrics by moving the population of common fields to record_compilation_metrics() instead of populating those independently wherever we use a the metrics_context contextmanager:
* Keep track of start and end time in MetricsContext and pass those to record_compilation_metrics() and populate those fields in that function.
* Pass exception info to record_compilation_metrics() and populate those field in that function.
* Add a new contextmanager, chromium_event_timed, to create the start/end "dynamo" event. This is important because I want this contextmanager to complete _after_ building the CompilationMetrics.
* Populate the compile_id field centrally in record_compilation_metrics().
* Populate the structured_logging_overhead centrally in record_compilation_metrics().
* Add the CompilationMetrics to the current chromium event in record_compilation_metrics(), after all common fields have been added. In a future diff, I can also add _all_ compilation metrics to the chromium event.

Test plan: Unit tests. Also see internal tlparse here: https://fburl.com/hv3inju2

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

masnesral commented on torch/_dynamo/convert_frame.py:
```diff
@@ -873,17 +870,11 @@ def count_args(code: CodeType) -> int:
-        "dynamo", chromium_start_time, {}, log_pt2_compile_event=True
-    )
-
     metrics_context = get_metrics_context()
```
@xmfan ; I guess there will be a merge conflict around here with your PR. You wanna land first?

xmfan replied:
should I still use chromium_event_log directly here? I don't need a context manager since my start and end times are pre-recorded

masnesral replied:
@xmfan yeah I think it's fine to just create your separate chromium event directly. I introduced the contextmanager here to help with the sequencing of the "dynamo" chromium event and the dynamo_compile scuba log (I need the scuba logging to complete first, then the chromium event).

masnesral replied:
I was hoping to let you in line first and I'd handle the merge, but I'm out for the week so I just landed. Sounds like Ed is voting for doing the compiled autograd logging earlier and potentially changing the compile_id format so maybe your PR moves the logging to another location anyway?

masnesral commented:
@pytorchbot merge

masnesral committed:
Update on "[logging] Move population of common MetricsContext fields to record_compilation_metrics"


Summary: Fix outstanding TODOs related to logging of CompilationMetrics by moving the population of common fields to record_compilation_metrics() instead of populating those independently wherever we use a the metrics_context contextmanager:
* Keep track of start and end time in MetricsContext and pass those to record_compilation_metrics() and populate those fields in that function.
* Pass exception info to record_compilation_metrics() and populate those field in that function.
* Add a new contextmanager, chromium_event_timed, to create the start/end "dynamo" event. This is important because I want this contextmanager to complete _after_ building the CompilationMetrics.
* Populate the compile_id field centrally in record_compilation_metrics().
* Populate the structured_logging_overhead centrally in record_compilation_metrics().
* Add the CompilationMetrics to the current chromium event in record_compilation_metrics(), after all common fields have been added. In a future diff, I can also add _all_ compilation metrics to the chromium event.

Test plan: Unit tests. Also see internal testing:
* dynamo_compile: https://fburl.com/scuba/dynamo_compile/sandbox/jrascnf9
* pt2_compile_events: https://fburl.com/scuba/pt2_compile_events/l3jnla06
* tlparse: https://fburl.com/bq5a9nqs

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

masnesral commented:
@pytorchbot merge
